src/librustdoc/html/format.rs

```diff
@@ -534,7 +535,7 @@ fn resolved_path(w: &mut fmt::Formatter, did: DefId, path: &clean::Path,
             } else {
                 format!("{}", HRef::new(did, &last.name))
             };
-            write!(w, "{}{}", path, last.params)?;
+            write!(w, "{}{}{}", if need_paren { "(" } else { "" }, path, last.params)?;
```

Comment:
This is missing the `print_all` and `w.alternate()` cases above although that won't be an issue if you render the parenthesis in `fmt_type` itself like I suggest.